pkgs/build-support/testers/expect-failure.sh

```diff
@@ -34,18 +34,20 @@ echo "testBuildFailure: Original builder produced exit code: $r"
 
 # -----------------------------------------
 # Write the build log to the default output
+
+# Source structured attrs and stdenv setup,
+# as per nixpkgs/pkgs/stdenv/generic/default-builder.sh
 #
-# # from stdenv setup.sh
-getAllOutputNames() {
-    if [ -n "$__structuredAttrs" ]; then
-        echo "${!outputs[*]}"
-    else
-        echo "$outputs"
-    fi
-}
+# NOTE: This MUST be done after the original builder has finished!
+#       Otherwise we could pollute its environment.
+if [ -e "${NIX_ATTRS_SH_FILE:-}" ]; then . "$NIX_ATTRS_SH_FILE"; elif [ -f .attrs.sh ]; then . .attrs.sh; fi
+source "$stdenv/setup"
```

Comment:
For now, could we copy the functionality we need?

Later, we could refactor `setup.sh` so that it is constructed at evaluation time with `builtins.toFile`. This also allows us to strip comments, so that we can improve those without mass rebuilds.